src/Core/Domain/CartRule/ValueObject/CartRuleAction.php

```diff
@@ -24,61 +24,60 @@
  * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
  */
 
-namespace PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\CartRuleAction;
+declare(strict_types=1);
 
-use PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\GiftProduct;
-use PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject\PercentageDiscount;
-use PrestaShop\PrestaShop\Core\Domain\ValueObject\Money;
+namespace PrestaShop\PrestaShop\Core\Domain\CartRule\ValueObject;
 
-/**
- * Cart rule action that gives free shipping.
- * It cannot have percentage or amount discount.
- * It can optionally have gift product.
- */
-final class FreeShippingAction implements CartRuleActionInterface
+use PrestaShop\PrestaShop\Core\Domain\CartRule\Exception\CartRuleConstraintException;
+
+class CartRuleAction
 {
     /**
-     * @var GiftProduct|null
+     * @var bool
      */
-    private $giftProduct;
+    private $freeShipping;
 
     /**
-     * @param GiftProduct|null $giftProduct
+     * @var GiftProduct|null
      */
-    public function __construct(?GiftProduct $giftProduct = null)
-    {
-        $this->giftProduct = $giftProduct;
-    }
+    private $giftProduct;
 
     /**
-     * {@inheritdoc}
+     * @var Discount|null
      */
-    public function isFreeShipping(): bool
-    {
-        return true;
-    }
+    private $discount;
 
-    /**
-     * {@inheritdoc}
-     */
-    public function getPercentageDiscount(): ?PercentageDiscount
-    {
-        return null;
+    public function __construct(
+        bool $freeShipping,
+        ?GiftProduct $giftProduct = null,
+        ?Discount $discount = null
+    ) {
+        if ($freeShipping || $giftProduct || $discount) {
+            $this->freeShipping = $freeShipping;
+            $this->giftProduct = $giftProduct;
+            $this->discount = $discount;
+
+            return;
+        }
+
+        throw new CartRuleConstraintException(
+            'Cart rule must have at least one action',
+            CartRuleConstraintException::MISSING_ACTION
+        );
```

Comment:
Early return feels weird in a constructor, I would have written it the other way around, but it's a detail!!

```suggestion
        if (!$freeShipping && !$giftProduct && $discount) {
            throw new CartRuleConstraintException(
                'Cart rule must have at least one action',
                CartRuleConstraintException::MISSING_ACTION
            );
        }

        $this->freeShipping = $freeShipping;
        $this->giftProduct = $giftProduct;
        $this->discount = $discount;
```

Reply:
yeah abit, but it is about the condition aswell. When using OR, it will not need to be executed further when it spots that freeShipping is set. While with && condition it would always have to check all the values. So even if it looks strange, I think we should just get used to it, because it seems to be the better way technically :thinking: